File: sdk-rust/src/models/ws.rs
use serde::{Deserialize, Serialize};

use super::{
    AccountOverview, Candlestick, MarketDepth, MarketPrice, MarketTrade,
    UserActiveTwap, UserFundingHistoryItem, UserOpenOrder, UserOrderHistoryItem,
    UserPosition, UserTradeHistoryItem,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WsMessage<T> {
    pub channel: String,
    pub data: T,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccountOverviewWsMessage {
    pub account_overview: AccountOverview,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserPositionsWsMessage {
    pub positions: Vec<UserPosition>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserOpenOrdersWsMessage {
    pub orders: Vec<UserOpenOrder>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserOrderHistoryWsMessage {
    pub orders: Vec<UserOrderHistoryItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserTradeHistoryWsMessage {
    pub trades: Vec<UserTradeHistoryItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserFundingHistoryWsMessage {
    pub funding: Vec<UserFundingHistoryItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketPriceWsMessage {
    #[serde(flatten)]
    pub price: MarketPrice,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AllMarketPricesWsMessage {
    pub prices: Vec<MarketPrice>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketTradesWsMessage {
    pub trades: Vec<MarketTrade>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CandlestickWsMessage {
    pub candle: Candlestick,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketDepthWsMessage {
    #[serde(flatten)]
    pub depth: MarketDepth,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserActiveTwapsWsMessage {
    pub twaps: Vec<UserActiveTwap>,
}

// --- WebSocket Subscribe/Unsubscribe Messages ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WsSubscribeRequest {
    pub method: String,
    pub subscription: String,
}

impl WsSubscribeRequest {
    pub fn subscribe(topic: &str) -> Self {
        Self {
            method: "subscribe".to_string(),
            subscription: topic.to_string(),
        }
    }

    pub fn unsubscribe(topic: &str) -> Self {
        Self {
            method: "unsubscribe".to_string(),
            subscription: topic.to_string(),
        }
    }
}

File: sdk-python/decibel_sdk/models/ws.py
"""WebSocket message wrappers for subscription topics."""

from __future__ import annotations

from typing import Generic, TypeVar

from pydantic import BaseModel

from decibel_sdk.models.account import (
    AccountOverview,
    UserFundingHistoryItem,
    UserTradeHistoryItem,
)
from decibel_sdk.models.market import (
    Candlestick,
    MarketDepth,
    MarketPrice,
    MarketTrade,
)
from decibel_sdk.models.order import (
    UserActiveTwap,
    UserOpenOrder,
    UserOrderHistoryItem,
)
from decibel_sdk.models.position import UserPosition

T = TypeVar("T")


class WsMessage(BaseModel, Generic[T]):
    channel: str
    data: T


class AccountOverviewWsMessage(BaseModel):
    account_overview: AccountOverview


class UserPositionsWsMessage(BaseModel):
    positions: list[UserPosition]


class UserOpenOrdersWsMessage(BaseModel):
    orders: list[UserOpenOrder]


class UserOrderHistoryWsMessage(BaseModel):
    orders: list[UserOrderHistoryItem]


class UserTradeHistoryWsMessage(BaseModel):
    trades: list[UserTradeHistoryItem]


class UserFundingHistoryWsMessage(BaseModel):
    funding: list[UserFundingHistoryItem]


class MarketPriceWsMessage(MarketPrice):
    pass


class AllMarketPricesWsMessage(BaseModel):
    prices: list[MarketPrice]


class MarketTradesWsMessage(BaseModel):
    trades: list[MarketTrade]


class CandlestickWsMessage(BaseModel):
    candle: Candlestick


class MarketDepthWsMessage(MarketDepth):
    pass


class UserActiveTwapsWsMessage(BaseModel):
    twaps: list[UserActiveTwap]


class WsSubscribeRequest(BaseModel):
    method: str
    subscription: str

    @classmethod
    def subscribe(cls, topic: str) -> WsSubscribeRequest:
        return cls(method="subscribe", subscription=topic)

    @classmethod
    def unsubscribe(cls, topic: str) -> WsSubscribeRequest:
        return cls(method="unsubscribe", subscription=topic)
